doc/man/man1/auks.1
.TH "auks" "1" "March 2009" "Matthieu Hautreux" "auks"

.SH "NAME"
auks \- client program to manage AUKS credentials

.SH "SYNOPSIS"
\fBauks\fR [\fI-h|--help\fR] [\fI-d|--debug\fR] [\fI-v|--verbose\fR] [\fI-f auks.conf|--config auks.conf\fR] [\fI-p|--ping\fR] [\fI-a|--add\fR] [\fI-g|--get\fR] [\fI-r|--remove\fR] [\fI-D|--dump\fR] [\fI-R (once|loop)|--renew (once|loop)\fR] [\fI-u uid|--uid uid\fR] [\fI-C ccache|--ccache\fR]

.SH "DESCRIPTION"
.LP
The AUKS commandline client can be used to manage AUKS daemon credentials.
Depending on his privileges, a client can use it to add/get/remove
credentials or dump all the available credentials.

.SH "OPTIONS"
.LP

.TP
\fB\-h, --help\fR
print this message

.TP
\fB\-v, --verbose\fR
increase verbosity output

.TP
\fB\-d, --debug\fR
increase debug output

.TP
\fB\-f auks.conf, --config auks.conf\fR
use \fIauks.conf\fR as the configuration file.
The default is \fB/etc/auks/auks.conf\fR.

.TP
\fB\-p, --ping\fR
do a simple ping request on the first responding AUKS daemon (default).
This operation requires at least to have a valid kerberos credential
to connect to the AUKS daemon

.TP
\fB\-a, --add\fR
add client kerberos credential to the first responding AUKS daemon. If 
client kerberos credential is not an addressless kerberos TGT, an
addressless cred is gotten and then transmitted in place of the given one.
This operation requires at least the \fBguest\fR privilege.

.TP
\fB\-g, --get\fR
get AUKS kerberos credential of the requested uid using the first responding 
AUKS daemon and stores it into the requested credential cache. This operation 
requires at least the \fBuser\fR privilege.

.TP
\fB\-r, --remove\fR
remove AUKS kerberos credential of the requested uid using the first responding 
AUKS daemon. This operation requires at least the \fBuser\fR privilege.

.TP
\fB\-D, --dump\fR
dump all the AUKS kerberos credentials using the first responding AUKS daemon.
This operation requires at least the \fBadmin\fR privilege.

.TP
\fB\-R (once|loop), --renew (once|loop)\fR
use AUKS commandline client as a credential renewer process. In \fBloop\fR mode,
the process exits only when an error occured or the credential can no longer 
be renewed. Used renewal mechanism is to first try to renew credential using 
the first responding AUKS daemon and second to use kerberos KDC in case of 
failure. In \fBonce\fR mode, the ticket is renewed once, only if needed, and 
the program exits. This operation requires at least the \fBuser\fR privilege.

.TP
\fB\-u uid, --uid uid\fR
specify the requested uid to use with get/remove requests

.TP
\fB\-C ccache, --ccache ccache\fR
specify the input/output credential file to use with add/get requests


.SH "ERRORS"
.LP
exit status is zero in case of success, not zero otherwise.

.SH "ENVIRONMENT VARIABLES"
.LP
.TP
\fBKRB5CCNAME\fR
Like any kerberos client, kerberos credentials used during AUKS
communications are stored into the default cache or the cache 
pointed	to by this environment variable.

.SH "COPYING"
.LP
Copyright  CEA/DAM/DIF (2009)
.br

.br
This software is a computer program whose purpose is to simplify
.br
the addition of kerberos credential support in Batch applications.
.br

.br
This software is governed by the CeCILL-C license under French law and
.br
abiding by the rules of distribution of free software.  You can  use, 
.br
modify and/ or redistribute the software under the terms of the 
.br
CeCILL-C license as circulated by CEA, CNRS and INRIA at the 
.br
following URL "http://www.cecill.info". 
.br

.br
As a counterpart to the access to the source code and  rights to copy,
.br
modify and redistribute granted by the license, users are provided only
.br
with a limited warranty  and the software's author,  the holder of the
.br
economic rights,  and the successive licensors  have only  limited
.br
liability. 
.br

.br
In this respect, the user's attention is drawn to the risks associated
.br
with loading,  using,  modifying and/or developing or reproducing the
.br
software by the user in light of its specific status of free software,
.br
that may mean  that it is complicated to manipulate,  and  that  also
.br
therefore means  that it is reserved for developers  and  experienced
.br
professionals having in-depth computer knowledge. Users are therefore
.br
encouraged to load and test the software's suitability as regards their
.br
requirements in conditions enabling the security of their systems 
.br
and/or data to be ensured and,  more generally, to use and operate 
.br
it in the same conditions as regards security. 
.br

.br
The fact that you are presently reading this means that you have had
.br
knowledge of the CeCILL-C license and that you accept its terms.
.br

.SH "SEE ALSO"
.BR auks.conf (5),
.BR auksd (8)



src/auks/auks.c
/***************************************************************************\
 * auks.c - user/admin interface to manage AUKS credentials 
 ***************************************************************************
 * Copyright  CEA/DAM/DIF (2009)
 *
 * Written by Matthieu Hautreux <matthieu.hautreux@cea.fr>
 * 
 * This software is a computer program whose purpose is to simplify
 * the addition of kerberos credential support in Batch applications.
 *
 * This software is governed by the CeCILL-C license under French law and
 * abiding by the rules of distribution of free software.  You can  use, 
 * modify and/ or redistribute the software under the terms of the CeCILL-C
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info". 
 * 
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability. 
 * 
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or 
 * data to be ensured and,  more generally, to use and operate it in the 
 * same conditions as regards security. 
 * 
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL-C license and that you accept its terms.
 ***************************************************************************
 * Copyright  CEA/DAM/DIF (2009)
 * 
 * Ecrit par Matthieu Hautreux <matthieu.hautreux@cea.fr>
 *
 * Ce logiciel est un programme informatique servant à faciliter l'ajout
 * du support des tickets Kerberos aux applications Batch.
 * 
 * Ce logiciel est régi par la licence CeCILL-C soumise au droit français et
 * respectant les principes de diffusion des logiciels libres. Vous pouvez
 * utiliser, modifier et/ou redistribuer ce programme sous les conditions
 * de la licence CeCILL-C telle que diffusée par le CEA, le CNRS et l'INRIA 
 * sur le site "http://www.cecill.info".
 * 
 * En contrepartie de l'accessibilité au code source et des droits de copie,
 * de modification et de redistribution accordés par cette licence, il n'est
 * offert aux utilisateurs qu'une garantie limitée.  Pour les mêmes raisons,
 * seule une responsabilité restreinte pèse sur l'auteur du programme,  le
 * titulaire des droits patrimoniaux et les concédants successifs.
 * 
 * A cet égard  l'attention de l'utilisateur est attirée sur les risques
 * associés au chargement,  à l'utilisation,  à la modification et/ou au
 * développement et à la reproduction du logiciel par l'utilisateur étant 
 * donné sa spécificité de logiciel libre, qui peut le rendre complexe à 
 * manipuler et qui le réserve donc à des développeurs et des professionnels
 * avertis possédant  des  connaissances  informatiques approfondies.  Les
 * utilisateurs sont donc invités à charger  et  tester  l'adéquation  du
 * logiciel à leurs besoins dans des conditions permettant d'assurer la
 * sécurité de leurs systèmes et ou de leurs données et, plus généralement, 
 * à l'utiliser et l'exploiter dans les mêmes conditions de sécurité. 
 * 
 * Le fait que vous puissiez accéder à cet en-tête signifie que vous avez 
 * pris connaissance de la licence CeCILL-C, et que vous en avez accepté les
 * termes. 
\***************************************************************************/
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <getopt.h>

#include <string.h>

#include <time.h>

#include "auks/auks_error.h"
#include "auks/auks_api.h"
#include "xternal/xlogger.h"

#define PING_REQUEST    1
#define ADD_REQUEST     2
#define GET_REQUEST     3
#define REMOVE_REQUEST  5
#define DUMP_REQUEST    6
#define RENEW_REQUEST   7
#define SEND_REQUEST    8
#define RECEIVE_REQUEST 9

#define RENEW_MODE_ONCE             1
#define RENEW_MODE_UNTIL_REMOVAL    2
#define RENEW_MODE_UNTIL_EXPIRATION 3

static void
_print_creds(auks_cred_t* creds, int creds_nb)
{
	char starttime[32], endtime[32], renew_till[32];
	struct tm *tm;
	int i;

	printf("%-10s %-34s %-24s %-24s %-24s\n",
		"UID",
		"PRINCIPAL",
		"START_TIME",
		"END_TIME",
		"RENEW_UNTIL");

	for ( i = 0 ; i < creds_nb ; i++ ) {
		if ((tm = localtime(&creds[i].info.starttime)) == NULL ||
		    strftime(starttime,sizeof(starttime),"%FT%T%z",tm) == 0)
			continue;
		if ((tm = localtime(&creds[i].info.endtime)) == NULL ||
		    strftime(endtime,sizeof(endtime),"%FT%T%z",tm) == 0)
			continue;
		if ((tm = localtime(&creds[i].info.renew_till)) == NULL ||
		    strftime(renew_till,sizeof(renew_till),"%FT%T%z",tm) == 0)
			continue;

		printf("%-10u %-34s %s %s %s\n",
			creds[i].info.uid,
			creds[i].info.principal,
			starttime,
			endtime,
			renew_till);
	}
}

int
main(int argc,char** argv)
{

	int fstatus = -1 ;

	int action=0;
	char* ccache=NULL;
	uid_t requested_uid;

	int debug_level=0;
	int verbose_level=0;
	char* conf_file_string;

	/* options processing variables */
	char* progname;
	char* optstring="dvhf:H:pagrDC:u:R:";
	char* short_options_desc="\nUsage : %s [-h|--help] [-d|--debug] [-f conffile|--config configfile] \
[-C ccache|--ccache ccache] [-p|--ping] [-a|--add] [-g|--get] [-r|--remove [-D|--dump] [-R (once|loop)|--renew (once|loop)] [-u uid|--uid uid] [--send] [--receive]\n\n";
	char* addon_options_desc="\
\t-h --help\t\tshow this message\n\
\t-d --debug\t\tincrease debug level\n\
\t-v --verbose\t\tincrease verbose level\n\
\t-f --config conffile\tConfiguration file\n\
\t-p --ping\t\tping request (default)\n\
\t-a --add\t\tadd request\n\
\t-g --get\t\tget request\n\
\t-D --dump\t\tdump request\n\
\t-r --remove\t\tremove request\n\
\t--send\t\tEncode a credential to stdout\n\
\t--receive\t\tDecode a credential from stdin\n\
\t-R --renew mode\t\trenew credential according to specified mode\n\
\t-C --ccache ccache\tConfiguration file\n\
\t-u --uid uid\t\tuid of requested cred owner (get request only)\n\n";
	static struct option long_options[] = {
		{"help", no_argument, 0, 'h'},
		{"debug", no_argument, 0, 'd'},
		{"verbose", no_argument, 0, 'v'},
		{"config", required_argument, 0, 'f'},
		{"ping", no_argument, 0, 'p'},
		{"add", no_argument, 0, 'a'},
		{"get", no_argument, 0, 'g'},
		{"dump", no_argument, 0, 'D'},
		{"remove", no_argument, 0, 'r'},
		{"renew", required_argument, 0, 'R'},
		{"ccache", required_argument, 0, 'C'},
		{"uid", required_argument, 0, 'u'},
		{"send", no_argument, 0, 'x'},
		{"receive", no_argument, 0, 'X'}
	};
	int option_index = 0;
	int option;

	auks_engine_t engine;

	FILE* logfile=NULL;
	FILE* debugfile=NULL;

	auks_cred_t* creds;
	int creds_nb;

	int renew_mode;

	/* set default requested uid */
	requested_uid=geteuid();

	/* get current program name */
	progname=rindex(argv[0],'/');
	if(progname==NULL)
		progname=argv[0];
	else
		progname++;

	conf_file_string=NULL;

	/* process options */
	while((option = getopt_long(argc,argv,optstring, long_options, &option_index)) != -1)
	{
		switch(option)
		{
		case 'v' :
			verbose_level++;
			break;
		case 'd' :
			debug_level++;
			break;
		case 'f' :
			conf_file_string=strdup(optarg);
			break;
		case 'C' :
			ccache=strdup(optarg);
			if(ccache==NULL){
				fprintf(stderr,"memory allocation error!");
				return 1;
			}
			break;
		case 'p' :
			action=PING_REQUEST;
			break;
		case 'a' :
			action=ADD_REQUEST;
			break;
		case 'g' :
			action=GET_REQUEST;
			break;
		case 'r' :
			action=REMOVE_REQUEST;
			break;
		case 'R' :
			action=RENEW_REQUEST;
			if(strncmp(optarg,"once",5)==0)
				renew_mode = AUKS_API_RENEW_ONCE ;
			else if(strncmp(optarg,"loop",5)==0)
				renew_mode = AUKS_API_RENEW_LOOP ;
			else {
				fprintf(stdout,short_options_desc,progname);
				fprintf(stdout,"%s\n",addon_options_desc);
				exit(1);
			}
			break;
		case 'D' :
			action=DUMP_REQUEST;
			break;
		case 'u' :
			requested_uid=(uid_t)atoi(optarg);
			break;
		case 'x':
			action = SEND_REQUEST;
			break;
		case 'X':
			action = RECEIVE_REQUEST;
			break;
		case 'h' :
		default :
			fprintf(stdout,short_options_desc,progname);
			fprintf(stdout,"%s\n",addon_options_desc);
			exit(0);
			break;
		}
	}

	/* set verbosity and debug level */
	xdebug_setmaxlevel(debug_level);
	xverbose_setmaxlevel(verbose_level);

	/* load configuration */
	fstatus = auks_api_init(&engine,conf_file_string);
	if ( fstatus != AUKS_SUCCESS ) {
		fprintf(stderr,"Auks API init failed : %s\n",
			auks_strerror(fstatus));
		goto exit;
	}

	if ( verbose_level > 0 && action != SEND_REQUEST )
		auks_api_set_logfile(&engine,"/dev/stdout");
	else if ( action == RENEW_REQUEST ) {
		if((strlen(engine.renewer_logfile) > 0) &&
		   engine.renewer_loglevel &&
		   (logfile = fopen(engine.renewer_logfile,"a+"))) {
			xverbose_setstream(logfile);
			xerror_setstream(logfile);
			xverbose_setmaxlevel(engine.renewer_loglevel);
			xerror_setmaxlevel(engine.renewer_loglevel);
		}
		else {
			xverbose_setmaxlevel(0);
			xerror_setmaxlevel(0);
		}
		if((strlen(engine.renewer_debugfile) > 0) &&
		   engine.renewer_debuglevel &&
		   (debugfile = fopen(engine.renewer_debugfile,"a+"))) {
			xdebug_setstream(debugfile);
			xdebug_setmaxlevel(engine.renewer_debuglevel);
		}
		else {
			xdebug_setmaxlevel(0);
		}
	}
	
	switch(action){

	case ADD_REQUEST :
		fstatus = auks_api_add_cred(&engine,
					    ccache);
		break;

	case GET_REQUEST :
		fstatus = auks_api_get_cred(&engine,
					    requested_uid,
					    ccache);
		break;

	case DUMP_REQUEST :
		fstatus = auks_api_dump(&engine,&creds,&creds_nb);
		if ( fstatus == AUKS_SUCCESS )
			_print_creds(creds,creds_nb);
		break;

	case REMOVE_REQUEST :
		fstatus = auks_api_remove_cred(&engine,
					       requested_uid);
		break;

	case RENEW_REQUEST :

		fstatus = auks_api_renew_cred(&engine,ccache,renew_mode);

		break;

	case SEND_REQUEST :
		fstatus = auks_api_send_cred(&engine, requested_uid);
		break;

	case RECEIVE_REQUEST :
		fstatus = auks_api_receive_cred(&engine, ccache);
		break;

	case PING_REQUEST :
	default :
		fstatus = auks_api_ping(&engine);
		break;

	}

	if ( fstatus != AUKS_SUCCESS ) {
		fprintf(stderr,"Auks API request failed : %s\n",
			auks_strerror(fstatus));
	}
	else if (action != SEND_REQUEST)
	{
		fprintf(stdout,"Auks API request succeed\n");
	}

	auks_api_close(&engine);

exit:
	if ( conf_file_string != NULL )
		free(conf_file_string);

	return fstatus;
}
